File: codechef/feb_longchallenge_div2/theatre.py
arr = [ [1,1,9,8], [2,3,10,8], [1,5,9,7], [5,1,2,2] ]
t_i = []
t_j = []
ans = []
i_i = [0,1,2,3]
pen = 0

for i in range(4):
    max = [0,0,0]
    if i not in t_i:
        for j in range(4):
            if j not in t_j:
                if max[0] < arr[i][j]:
                    max[0], max[1], max[2] = arr[i][j], i, j
        for k in range(4):
            if k not in t_i:
                if max[0] < arr[k][max[2]]:
                    max[0], max[1] = arr[k][max[2]], k
        for l in range(4):
            if l not in t_j:
                if max[0] < arr[max[1]][l]:
                    max[0], max[2] = arr[max[1]][l], l

        ans.append(max[0])
        t_i.append(max[1])
        t_j.append(max[2])
        i_i.remove(max[1])

for i in i_i:
    max = [0,0,0]
    for j in range(4):
        if j not in t_j:
            if max[0]<arr[i][j]:
                max[0], max[1], max[2] = arr[i][j], i, j

    if max[0] is not 0:
        ans.append(max[0])
        t_i.append(max[1])
        t_j.append(max[2])
        i_i.remove(max[1])
    else:
        pen += 1

print(ans)
for i,j in zip(t_i, t_j):
    print(i,j)

print("pen", pen)

File: codechef/feb_longchallenge_div2/theatre.cpp
#include <iostream>
#include <map>
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

bool sortinrev(const pair<int,int> &a,  
               const pair<int,int> &b) 
{ 
       return (a.first > b.first); 
} 

int main(){
    int p[] = {100, 75, 50, 25};
    map<int, int> s = { make_pair(12,0),make_pair(3,1),make_pair(6,2),make_pair(9,3) };
    map<char, int> m = { make_pair('A',0),make_pair('B',1),make_pair('C',2),make_pair('D',3) };

    int t,n,st,pen;
    long long int score=0;
    char mv;

    cin >> t;
    while(t--){
        cin >> n;

        pen = 0;
        long long int l_score = 0;
        vector<vector<int>> m_t(4, vector<int> (4, 0));
        vector< pair<int, int> > max_r;
        /* t_m.clear(); */
        /* max_r.clear(); */

        for (int i=0; i<n; i++){
            cin >> mv >> st;
            m_t[s[st]][m[mv]] += 1;
        }

        /* cout << '\n'; */
        /* for(int i=0; i<4; i++){ */
        /*     for(int j=0; j<4; j++){ */
        /*         cout << m_t[i][j] << '\t'; */
        /*     } */
        /*     cout << '\n'; */
        /* } */

        vector<int> t_j;
        vector<int> t_i;
        vector<int> ii{0,1,2,3};
        for (int i=0; i<4; i++){
            /* pair <int,int> max(0,0); */
            if (find(t_i.begin(), t_i.end(), i) != t_i.end()) continue;
            int max[] = {0,0,0};
            for(int j=0; j<4; j++){
                if (find(t_j.begin(), t_j.end(), j) != t_j.end()) continue;
                if ( ((m_t[i][j]) > max[0]) ){
                    max[0] = m_t[i][j];
                    /* max.second = i; */
                    max[1] = i;
                    max[2] = j;
                }
            }
            for(int k=0; k<4; k++){
                if (find(t_i.begin(), t_i.end(), k) != t_i.end()) continue;
                if ( ((m_t[k][max[2]]) > max[0]) ){
                    max[0] = m_t[k][max[2]];
                    /* max.second = i; */
                    max[1] = k;
                }
            }
            for(int l=0; l<4; l++){
                if (find(t_j.begin(), t_j.end(), l) != t_j.end()) continue;
                if ( ((m_t[max[1]][l]) > max[0]) ){
                    max[0] = m_t[1][l];
                    /* max.second = i; */
                    max[2] = l;
                }
            }

            if (max[0] != 0){
                max_r.push_back(make_pair(max[0], 0));
                t_i.push_back(max[1]);
                t_j.push_back(max[2]);
                ii.erase(remove(ii.begin(), ii.end(), max[1]), ii.end());
            }

        }

        for (auto i: ii){
            int max[] = {0,0,0};
            for (int j=0; j<4; j++){
                if (find(t_j.begin(), t_j.end(), j) != t_j.end()) continue;
                if( max[0] < m_t[i][j] ){
                    max[0] = m_t[i][j];
                    max[1] = i;
                    max[2] = j;
                }
            }

            if (max[0] != 0){
                t_i.push_back(max[1]);
                t_j.push_back(max[2]);
                max_r.push_back(make_pair(max[0], 0));
            }else{
                pen += 1;
            }
        }

        /* cout << '~' << endl; */
        /* for (auto i: max_r){ */
        /*     cout << i.first << '\t'; */
        /* } */
        /* cout << '\n'; */
        /* for (auto i: t_i){ */
        /*     cout << i << '\t'; */
        /* } */
        /* cout << '\n'; */
        /* for (auto j: t_j){ */
        /*     cout << j << '\t'; */
        /* } */

        sort(max_r.begin(), max_r.end(), sortinrev);

        int c = 0;
        for (auto i: max_r){
            l_score += p[c] * i.first;
            c += 1;
        }

        l_score -= (100*pen);
        cout << l_score << endl;
        score += l_score;

    }

    cout << score << endl;

}
